### PR TITLE
build: Make G_FUZZING constexpr, require -DBUILD_FOR_FUZZING=ON to fuzz

### LINKED ISSUE #31178: bench: `linearizeoptimallyexample11` benchmark now running 4x slow than previously
According to https://corecheck.dev/benchmarks, the `linearizeoptimallyexample11` benchmark 4x'd it's execution time at some point in the last few days.
Screenshot 2024-10-29 at 14 15 02

### PR DESCRIPTION
**Pull request by maflcko**
`g_fuzzing` is used inside `Assume` at runtime, causing significant overhead in hot paths. See https://github.com/bitcoin/bitcoin/issues/31178

One could simply remove the `g_fuzzing` check from the `Assume`, but this would make fuzzing a bit less useful. Also, it would be unclear if `g_fuzzing` adds a runtime overhead in other code paths today or in the future.

Fix all issues by making `G_FUZZING` equal to the build option `BUILD_FOR_FUZZING`, and for consistency in fuzzing, require it to be set when executing any fuzz target.

Fixes https://github.com/bitcoin/bitcoin/issues/31178

Temporarily this drops fuzzing from two CI tasks, but they can be re-added in a follow-up with something like https://github.com/bitcoin/bitcoin/pull/31073

### CODE REVIEW

File: src/util/check.h [hunk 1]
```diff
@@ -44,7 +50,7 @@ void assertion_fail(std::string_view file, int line, std::string_view func, std:
 template <bool IS_ASSERT, typename T>
 constexpr T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] const char* file, [[maybe_unused]] int line, [[maybe_unused]] const char* func, [[maybe_unused]] const char* assertion)
 {
-    if (IS_ASSERT || std::is_constant_evaluated() || g_fuzzing
+    if (IS_ASSERT || std::is_constant_evaluated() || G_FUZZING
```
Comment (dergoegge) [inline src/util/check.h:53]:
```suggestion
    if constexpr (IS_ASSERT || std::is_constant_evaluated() || G_FUZZING
```

**Review by dergoegge — APPROVED** [review]
utACK fafbf8acf419d5e2ca307e5804099361ca7471af

File: src/util/check.h (hunk 1 above)
Comment (marcofleon) [reply src/util/check.h]: Do we need the `std::is_constant_evaluated()` then? I'm getting this warning everywhere when I build
```
../../../../src/util/check.h:53:32: warning: 'std::is_constant_evaluated' will always evaluate to 'true' in a manifestly constant-evaluated expression [-Wconstant-evaluated]
   53 |     if constexpr (IS_ASSERT || std::is_constant_evaluated() || G_FUZZING
      |                                ^
../../../../src/util/check.h:53:32: warning: 'std::is_constant_evaluated' will always evaluate to 'true' in a manifestly constant-evaluated expression [-Wconstant-evaluated]
```

**Comment by marcofleon** [comment]
Tested ACK fafbf8acf419d5e2ca307e5804099361ca7471af

**Comment by ryanofsky** [comment]
Code review ACK fafbf8acf419d5e2ca307e5804099361ca7471af but approach -0, because this approach means libraries built for fuzz testing do not function correctly if used in a release, and libraries built for releases are mostly useless for fuzz testing. So I would like to at least consider other solutions to this problem even if we go with this one.

#31178 makes it pretty clear that if we want to be able to write `Assume()` statements in hot paths, we need to be able to compile them out in release builds to avoid impacting performance, and leave them compiled into builds that are used for fuzzing. So I do think the build system should support compiling libraries differently for fuzz testing and releases. But I don't think the build system should go so far as to make libraries built for fuzz testing and libraries built for releases mutually incompatible.

Also I don't really buy the idea that if you put an `Assume()` statement in a hot path, you just be to assume it has no cost and that compiler will optimize it out. I think majority of Assume() statements are not in hot paths, and main use-case for Assume() is not to be a faster Assert() that you use in performance-critical code, but to be a safer Assert() that you can use to catch bugs during development, but will not crash the entire program and provide a terrible user experience when bugs (inevitably) occur in production.

So I think a better alternative to this PR might just be to provide a better alternative to Assume(). I think most uses of Assume() are fine as they are, but a few really are on hot paths, and I tried adding a simple counter to identify them in the linearizeoptimallyexample11 benchmark. If I disable these checks, the speedup is equivalent to this PR without drawbacks of this PR. The change I would propose based on this is:

<details><summary>diff</summary>
<p>

```diff
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -225,6 +225,7 @@ if(BUILD_FOR_FUZZING)
   set(BUILD_GUI_TESTS OFF)
   set(BUILD_BENCH OFF)
   set(BUILD_FUZZ_BINARY ON)
+  target_compile_definitions(core_interface INTERFACE ENABLE_SLOWCHECK)
 endif()
 
 include(ProcessConfigurations)
--- a/cmake/module/ProcessConfigurations.cmake
+++ b/cmake/module/ProcessConfigurations.cmake
@@ -126,6 +126,7 @@ target_compile_definitions(core_interface_debug INTERFACE
   DEBUG_LOCKCONTENTION
   RPC_DOC_CHECK
   ABORT_ON_FAILED_ASSUME
+  ENABLE_SLOWCHECK
 )
 # We leave assertions on.
 if(MSVC)
--- a/src/cluster_linearize.h
+++ b/src/cluster_linearize.h
@@ -337,7 +337,7 @@ struct SetInfo
     /** Add a transaction to this SetInfo (which must not yet be in it). */
     void Set(const DepGraph<SetType>& depgraph, ClusterIndex pos) noexcept
     {
-        Assume(!transactions[pos]);
+        SLOWCHECK(!transactions[pos]);
         transactions.Set(pos);
         feerate += depgraph.FeeRate(pos);
     }
--- a/src/util/bitset.h
+++ b/src/util/bitset.h
@@ -281,7 +281,7 @@ class MultiIntBitSet
         /** Progress to the next 1 bit (only if != IteratorEnd). */
         constexpr Iterator& operator++() noexcept
         {
-            Assume(m_idx < N);
+            SLOWCHECK(m_idx < N);
             m_val &= m_val - I{1U};
             if (m_val == 0) {
                 while (true) {
@@ -301,7 +301,7 @@ class MultiIntBitSet
         /** Get the current bit position (only if != IteratorEnd). */
         constexpr unsigned operator*() const noexcept
         {
-            Assume(m_idx < N);
+            SLOWCHECK(m_idx < N);
             return m_pos;
         }
     };
@@ -316,7 +316,7 @@ public:
     /** Set a bit to 1. */
     void constexpr Set(unsigned pos) noexcept
     {
-        Assume(pos < MAX_SIZE);
+        SLOWCHECK(pos < MAX_SIZE);
         m_val[pos / LIMB_BITS] |= I{1U} << (pos % LIMB_BITS);
     }
     /** Set a bit to the specified value. */
--- a/src/util/check.h
+++ b/src/util/check.h
@@ -81,6 +81,12 @@ constexpr T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] con
 /**
  * Assume is the identity function.
  *
+ * Assume() should be used instead of Assert() in cases where if the condition
+ * is not true, it indicates there is a bug, and you want to handle the bug by
+ * crashing in debug builds and in fuzz tests, but want to avoid unnecessary
+ * crashes in release builds and handle it with logging, warning, or other
+ * fallback behavior.
+ *
  * - Should be used to run non-fatal checks. In debug builds it behaves like
  *   Assert()/assert() to notify developers and testers about non-fatal errors.
  *   In production it doesn't warn or log anything.
@@ -90,6 +96,24 @@ constexpr T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] con
  */
 #define Assume(val) inline_assertion_check<false>(val, __FILE__, __LINE__, __func__, #val)
 
+/**
+ * SLOWCHECK() can be used to perform checks that are enabled in debug and fuzz
+ * builds but are skipped in release builds. It is meant to be used for checks
+ * that are slow, either because they occur in hot code paths, or because the
+ * checks themselves are expensive. Assert() should be preferred for critical
+ * checks which are not in performance-sensitive code.
+ *
+ * SLOWCHECK() is basically equivalent to the traditional C/C++ assert() macro
+ * For historic reasons, Bitcoin Core cannot be compiled with NDEBUG so there is
+ * no way to skip assert() checks in release builds, and SLOWCHECK() restores
+ * this lost functionality.
+ */
+#ifdef ENABLE_SLOWCHECK
+#define SLOWCHECK(val) assert(val);
+#else
+#define SLOWCHECK(val) assert(1 || (val));
+#endif
+
 /**
  * NONFATAL_UNREACHABLE() is a macro that is used to mark unreachable code. It throws a NonFatalCheckError.
  */
--- a/src/util/vecdeque.h
+++ b/src/util/vecdeque.h
@@ -74,7 +74,7 @@ class VecDeque
     /** What index in the buffer does logical entry number pos have? */
     size_t BufferIndex(size_t pos) const noexcept
     {
-        Assume(pos < m_capacity);
+        SLOWCHECK(pos < m_capacity);
         // The expression below is used instead of the more obvious (pos + m_offset >= m_capacity),
         // because the addition there could in theory overflow with very large deques.
         if (pos >= m_capacity - m_offset) {
```
</p>
</details>

I would curious to know from @sipa and others if they think this approach makes sense and does not add too much of a burden.

Specifically results I saw testing this with the linearizeoptimallyexample11 benchmark were 1.42 op/sec on master, 1.77 op/sec reverting 9f243cd7fa6654e3b71ba6bff82cceed547c5d53 (#31093), 1.77 op/sec cherry-picking fafbf8acf419d5e2ca307e5804099361ca7471af from this PR, and 1.83 op/sec with the proposed change.

**Comment by ryanofsky** [comment]
> I don't think it is a supported use-case to take libraries from one build (with different build options) and drop them into another build

Agree it should not be a generally supported use case. And I do think it's nice to have a BUILD_FOR_FUZZING option that enables the best options for builds specifically doing fuzzing.

But for normal builds, I don't think the whole codebase should need to be recompiled with different options just to get a useful fuzz binary. And I don't think it is good for BUILD_FOR_FUZZING option to create strange libraries that seem functional but skip proof of work checks. I did ACK this PR in case we think these problems are worth having so `Assume()` can be optimized out in hot code paths. But I think these are negative consequences of this PR, even if worth the tradeoff.

> > Also I don't really buy the idea that if you put an Assume() statement in a hot path, you just be to assume it has no cost and that compiler will optimize it out.
>
> I don't think I've claimed this.

Yes sorry, the comments above about Assume() usage were meant for @sipa who wrote "I have been using Assume in many places assuming it'll be optimized out entirely in production code" and similar things in the other issues.

> It may be best to submit this as a separate pull request.

Yes, will do if it seems like something that is helpful.

**Comment by davidgumberg** [comment]
> * Everything in an `if constexpr` is still looked at by the compiler and any code issues should be detected.

Didn't think of this, that makes sense to me. I also may have misunderstood or misrepresented the use-case/concern I tried to describe, but it's not my own and I don't have any view on it, so it seems silly that I even mentioned it.

I still ACK https://github.com/bitcoin/bitcoin/pull/31191/commits/fafbf8acf419d5e2ca307e5804099361ca7471af for fixing the regression measured in #31178.

**Comment by ryanofsky** [comment]
The use-case which I think this breaks and should be supported is the ability to use fuzzing in a normal build instead of a dedicated build.

As an analogy, if you want use a debugger, the best place to use it is in a dedicated debug build, but you should also be able to generate debug symbols and use debuggers with some limitations in release builds. Or if you want to check memory safety, the best way to do it may be to run MSan or valgrind in dedicated builds, but it should also be possible to use ASan in normal builds. Similarly, if you want to run fuzz tests, the best way to run them is in a dedicated build with -DBUILD_FOR_FUZZING=ON, but I think you should also be able to run fuzz tests in a normal build with other functioning executables. This approach lets you have a single build supporting most developer tools, even if it is not the ideal build for running those tools, instead of separate builds for every tool. I think it lowers the barrier to start trying new tools if it only requires toggling build options instead of creating entirely new build configurations.

AFAICT before #31093, libraries with BUILD_FOR_FUZZING did not function normally because they skipped proof of work checks, and libraries without BUILD_FOR_FUZZING were not very useful for fuzzing because they skipped most Assume() checks. #31093 fixed both of those issues and this PR breaks them again. Maybe this is acceptable, but doesn't seem good, and an alternate approach like the one in https://github.com/bitcoin/bitcoin/pull/31191#issuecomment-2451865832 might be better. I'd be happy to open a PR with that approach it makes sense, or to learn that I am wrong and it doesn't make sense, or to get feedback that it is not a good solution for other reasons.

I also think this PR is ok, but I don't like the tradeoff it is making, so I'm just pointing out the disadvantages I see here.

**Comment by marcofleon** [comment]
> I think it lowers the barrier to start trying new tools if using them only requires toggling build options not creating entirely new build configurations.

I see what you're saying about tools being more accessible if they're able to be used within a single build. However, I do think with cmake it's straightforward to maintain separate builds for different purposes. And fuzz testing might be involved enough to warrant its own build.

I don't see many advantages to fuzzing in a normal build that isn't optimized for it. If we're still able to verify that the fuzz binary builds properly and run pre-fuzz checks, as mentioned in https://github.com/bitcoin/bitcoin/pull/31191#issuecomment-2451410917, then to me it feels reasonable to require BUILD_FOR_FUZZING when we intend to do some actual fuzz testing.

I also think it's cleaner to isolate fuzz-specific code paths (like skipping the PoW check) to a dedicated build. Of course, I'm open to reviewing other potential solutions.

**Comment by ryanofsky** [comment]
Went ahead and merged this since it has enough ACKs, and it sounds like the concerns I have about drawbacks of this PR don't seem to be shared by the other reviewers. I think it would be nice if fuzzing could be turned on without turning every else off, and this approach will probably make me a little less likely to start writing fuzz tests, and prefer writing unit tests instead for convenience. But status after this PR isn't worse than the status before #31093, and this approach can always be revisited later.

**Comment by ryanofsky** [comment]
> I understand your concerns, but did you actually encounter them in reality?

Yes, I've never written or run a fuzz test because they require separate setup and because I'm reasonably happy writing unit tests. If I could run fuzz tests in my default build, I'd be more likely to write fuzz tests as an alternative to unit tests. I understand fuzz tests are not always an alternative to unit tests, but I've written a lot of unit tests where I'm just handcrafting inputs to test edge cases, and then checking invariants, and probably fuzz tests could save effort and do a better job. Not just letting developers add just -fsanitize=fuzzer,address to their flags and run an extra binary seems like adding an unnecessary obstacle to starting to use the fuzzer. Maybe the obstacle has to exist for reasons I'm too uninformed to know about, but from the outside treating fuzzing compiler options differently than other compiler option seems confusing and unnecessarily complex.

**Comment by ryanofsky** [comment]
Thanks for clarifications and ideas. Iit sounds like doing what I want would require `fuzzer-no-link` instead of `fuzzer` but otherwise it should be possible for BUILD_FOR_FUZZING to just enable extra instrumentation and not make every library and every other binary unusable. To me this just seems like a most straightforward way and least surprising way of organizing the build, that would make fuzzing easier to get started with by just being able to turn on an option that doesn't break everything else. I can see why this approach is working for people, though, even if it seems not very cleanly designed.

**Pull request merged**